src/linear_algebra/general_linear_group.lean

```diff
@@ -0,0 +1,200 @@
+units.map_equiv to_lin_alg_equiv'.to_mul_equiv
+
+/--Given a matrix with invertible determinant we get an element of `GL n R`-/
+ def GL.mk' (A : matrix n n R) (h : invertible (det A)): GL n R:=
+unit_of_det_invertible A
+
+/--Given a matrix with unit determinant we get an element of `GL n R`-/
+noncomputable def GL.mk'' (A : matrix n n R) (h: is_unit (det A)): GL n R:=
+nonsing_inv_unit A h
+
+end
+namespace general_linear_group
+
+variables {n : Type u} [decidable_eq n] [fintype n] {R : Type v} [comm_ring R]
+
+instance coe_fun : has_coe_to_fun (GL n R) :=
+{ F   := λ _, n → n → R,
+  coe := λ A, A.val }
+
+/--Given a element of `GL n R` it gives the associated linear map-/
+
+def to_lin' (A : GL n R) := matrix.to_lin' A
+
+lemma ext_iff (A B : GL n R) : A = B ↔ (∀ i j, A i j = B i j) :=
+begin
+  dsimp at *, fsplit, work_on_goal 0 { intros a i j, induction a, refl }, intros a,
+  ext1, dsimp at *, ext1, solve_by_elim,
+end
+
+@[ext] lemma ext (A B : GL n R) : (∀ i j, A i j = B i j) → A = B :=
+(ext_iff A B).mpr
+
+/-- Applying `matrix.nonsing_inv` to the coercion is the same as coercing the group inverse. -/
+lemma nonsing_inv_coe_eq_coe_inv (A : GL n R) : (↑A : matrix n n R)⁻¹ = ↑(A⁻¹) :=
+begin
+  letI := A.invertible,
+  exact (inv_eq_nonsing_inv_of_invertible ↑A).symm,
+end
+
+section coe_lemmas
+
+variables (A B : GL n R)
+
+@[simp] lemma mul_val : ↑(A * B) = A ⬝ B := rfl
+
+@[simp] lemma mul_apply : ⇑(A * B) = (A ⬝ B) := rfl
+
+@[simp] lemma one_val : ↑(1 : GL n R) = (1 : matrix n n R) := rfl
+
+@[simp] lemma one_apply : ⇑(1 : GL n R) = (1 : matrix n n R) := rfl
+
+@[simp] lemma inv_val : ↑(A⁻¹) = nonsing_inv A :=
+begin
+  have:=nonsing_inv_coe_eq_coe_inv A, unfold_coes at *, rw ← this, refl,
+end
+
+@[simp] lemma inv_apply : ⇑(A⁻¹) = nonsing_inv A :=
+begin
+  have:=nonsing_inv_coe_eq_coe_inv A, unfold_coes at *, rw ← this, refl,
+end
+
+@[simp] lemma to_lin'_mul :to_lin' (A * B) = (to_lin' A).comp (to_lin' B) :=
+matrix.to_lin'_mul A B
+
+@[simp] lemma to_lin'_one :to_lin' (1 : GL n R) = linear_map.id :=
+matrix.to_lin'_one
+
+end coe_lemmas
+
+instance SL_to_GL : has_coe (special_linear_group n R) (GL n R):=
+⟨λ A, ⟨↑A, ↑(A⁻¹), congr_arg coe (mul_right_inv A), congr_arg coe (mul_left_inv A)⟩⟩
+
+@[simp] lemma GL_vals (A : GL n R): ∀ i j, A i j = A.1 i j:=
+begin
+unfold_coes, intros i j,refl,
+end
+
+lemma det_not_zero [nontrivial R] (A : GL n R): det A ≠ 0:=
+begin
+  have:=GL.is_unit_det _ _ A, simp, by_contradiction, unfold_coes at *,
+  rw h at this, simp at *, exact this,
+end
+
+end general_linear_group
+
+namespace GL_plus
+
+variables  (n : Type u) [decidable_eq n] [fintype n] (R : Type v) [linear_ordered_comm_ring R ]
+
+lemma one_in_GL_pos : 0 < det (1 : GL n R ) :=
+begin
+simp only [det_one, gt_iff_lt, general_linear_group.one_apply,zero_lt_one],
+end
+
+lemma mul_in_GL_pos (A B : GL n R ) (h1: 0 <  det A  ) (h2: 0 <  det B ): 0 < det (A*B):=
+
+begin
+simp only [gt_iff_lt, det_mul, mul_eq_mul], apply mul_pos h1 h2,
+end
```

Comment:
Some more style changes which apply elsewhere in this file:
```suggestion
lemma mul_in_GL_pos (A B : GL n R) (h1: 0 < det A) (h2 : 0 < det B) : 0 < det (A*B) :=
begin
  simp only [gt_iff_lt, det_mul, mul_eq_mul], apply mul_pos h1 h2,
end
```
In particular: spaces around colons, no spaces around the the content of `()` or `[]`, avoid using two spaces in a row, and don't leave a blank line between a lemma and its proof.

Reply:
OK I've these things now. I hope I got them all. Is there some automatic way to check for these things?